[ROCm] Faster Custom Paged Attention kernels

tjtanaa opened the pull request:
## Description
This PR implements a faster Custom Paged Attention (CPA) kernel based on `mfma16x16x16` instructions.
This feature is from ROCm/vllm (https://github.com/ROCm/vllm/pull/372).


## End-to-End Performance gain
Model: Llama-3.1-70B-Instruct
Tensor Parallelism: 1
GPU: MI300X

| CPA Version | Input length | Output length | KV-cache-dtype | Quantization | Prompt numbers | Req/s | Total Tokens/s | Output Tokens/s |
|-------------|--------------|---------------|----------------|--------------|----------------|-------|----------------|-----------------|
| before changes | 128          | 128           | fp8_e4m3       | fp8          | 200            | 13.05 | 3340.6         | 1670.3          |
| before changes | 128          | 256           | fp8_e4m3       | fp8          | 200            | 7.56  | 2901.31        | 1934.21         |
| before changes | 128          | 2048          | fp8_e4m3       | fp8          | 200            | 0.78  | 1698.35        | 1598.45         |
| before changes | 512          | 128           | fp8_e4m3       | fp8          | 200            | 6.44  | 4122.57        | 824.51          |
| before changes | 512          | 256           | fp8_e4m3       | fp8          | 200            | 4.48  | 3443.46        | 1147.82         |
| before changes | 512          | 2048          | fp8_e4m3       | fp8          | 200            | 0.66  | 1696.64        | 1357.31         |
| before changes | ShareGPT |               | fp8_e4m3       | fp8          | 1000           | 6.22  | 2574.19        | 1234.64         |
| optimized   | 128          | 128           | fp8_e4m3       | fp8          | 200            | 15.11 | 3867.75        | 1933.87         |
| optimized   | 128          | 256           | fp8_e4m3       | fp8          | 200            | 9.01  | 3459.98        | 2306.65         |
| optimized   | 128          | 2048          | fp8_e4m3       | fp8          | 200            | 1.2   | 2609.04        | 2455.57         |
| optimized   | 512          | 128           | fp8_e4m3       | fp8          | 200            | 7.33  | 4694.05        | 938.81          |
| optimized   | 512          | 256           | fp8_e4m3       | fp8          | 200            | 5.5   | 4223.29        | 1407.76         |
| optimized   | 512          | 2048          | fp8_e4m3       | fp8          | 200            | 1.03  | 2648.55        | 2118.84         |
| optimized   | ShareGPT |               | fp8_e4m3       | fp8          | 1000           | 7.45  | 3081.14        | 1477.79         |

vllmellm committed:
added rocm custom paged attention.

Ported ROCm/vllm changes to upstream vLLM

This commit manually ports changes from ROCm/vllm (https://github.com/ROCm/vllm/pull/372) to upstream vLLM.
The original work was done by sanyalington.

Co-authored-by: sanyalington <shomy.sanyal@amd.com>

Signed-off-by: vllmellm <vllm.ellm@embeddedllm.com>

tjtanaa committed:
format code and enable test_attention.py in AMD CI

Signed-off-by: tjtanaa <tunjian.tan@embeddedllm.com>

sanyalington committed:
add author; update requirements-rocm.txt

Signed-off-by: sanyalington <shomy.sanyal@amd.com>

vllmellm committed:
[Misc]: removed unnecessary condition in atttention test.

tjtanaa commented:
Regarding to the API changes of `paged_attention` in `csrc/rocm/torch_bindings.cpp`. This change only affects ROCm code path and does not interfere with code path of other platform.
```
 rocm_ops.def(
      "paged_attention(Tensor! out, Tensor exp_sums,"
      "                Tensor max_logits, Tensor tmp_out,"
      "                Tensor query, Tensor key_cache,"
      "                Tensor value_cache, int num_kv_heads,"
      "                float scale, Tensor block_tables,"
      "                Tensor context_lens, int block_size,"
      "                int max_context_len,"
      "                Tensor? alibi_slopes,"
      "                str kv_cache_dtype,"
      "                float k_scale, float v_scale,"
      "                Tensor? fp8_out_scale,"
      "                int partition_size) -> ()");
 ```
 
 ### Seeking advice on handling the variables `fp8_out_scale` and `partition_size`.
 **Situation:** Currently these two variables `fp8_out_scale` and `partition_size` has been introduced in the Custom Paged Attention ROCm, but they are not in used by higher level abstractions. They are set to `fp8_out_scale=None` and `partition_size=256`. The `partition_size=256` has been found experimentally to be a good value for MI300.
 
**Option 1:** 
* Remove `fp8_out_scale` from `csrc/rocm/attention.cu`
* Hard code partition_size to be 256 in `csrc/rocm/attention.cu`.
This avoid changing the `paged_attention` API in `csrc/rocm/torch_bindings.cpp`

**Option 2:**
* Keep the variables as is, and mark *TODO:* for future feature update to remember introducing fp8 scaling strategy for ROCm.
* Set `fp8_out_scale=None` and `partition_size=256` when calling `ops.paged_attention_rocm` in `vllm/attention/backends/rocm_flash_attn.py`

hongxiayang commented:
@tjtanaa  Please fix the DCO error: 
Ensure you have a local copy of your branch by [checking out the pull request locally via command line](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/checking-out-pull-requests-locally).
In your local branch, run: git rebase HEAD~4 --signoff
Force push your changes to overwrite the branch: git push --force-with-lease origin port-rocm-cpa-credit

vllmellm committed:
[Kernel][Hardware][AMD] refactoring rocm custom paged attention to eliminate the need for  additional argumnets (partition_size and fp8_output_scale) in its api.

Signed-off-by: vllmellm <vllm.ellm@embeddedllm.com>

vllmellm committed:
Merge remote-tracking branch 'origin/main' into port-rocm-cpa-credit

Signed-off-by: vllmellm <vllm.ellm@embeddedllm.com>

vllmellm committed:
[kernel] fix the format.

Signed-off-by: vllmellm <vllm.ellm@embeddedllm.com>

vllmellm committed:
[Kernel][Hardware][AMD] improved rocm custom paged attention accuracy and code documentation. updated its unittest to match the correct partition size based on paged attention versions as well as platform type.

Signed-off-by: vllmellm <vllm.ellm@embeddedllm.com>

tjtanaa commented:
> @tjtanaa Please fix the DCO error: Ensure you have a local copy of your branch by [checking out the pull request locally via command line](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/checking-out-pull-requests-locally). In your local branch, run: git rebase HEAD~4 --signoff Force push your changes to overwrite the branch: git push --force-with-lease origin port-rocm-cpa-credit

@hongxiayang We find that rebasing is hard as we had merged from `main`. In the process of fixing the DCO, we had to resolve merge conflict twice, and will require us to test everything again. It seems there are ways to override the DCO during merge. Could we get more input from vLLM maintainers about DCO issue.

tjtanaa committed:
Merge remote-tracking branch 'origin/main' into port-rocm-cpa-credit

Signed-off-by: tjtanaa <tunjian.tan@embeddedllm.com>

tjtanaa committed:
format and lint code

Signed-off-by: tjtanaa <tunjian.tan@embeddedllm.com>

hongxiayang reviewed (approved):
Thanks a lot. LGTM.

hongxiayang commented:
I also verified the throughput numbers using the image built from @tjtanaa 's branch.

hongxiayang commented:
@tjtanaa  Can you work with @DarkLight1337 to see what else is needed in order to merge this PR?
Thanks for your effort for upstreaming this and fixing the test and clean up other spelling errors as well.

DarkLight1337 commented:
@tlrmchlsmth @WoosukKwon can either of you take a look at this PR?

tlrmchlsmth commented on tests/kernels/test_attention.py:
```diff
@@ -144,6 +145,8 @@ def test_paged_attention(
             or (version == "rocm" and head_size not in (64, 128))):
         pytest.skip()
 
+    global PARTITION_SIZE
+
```
why make PARTITION_SIZE a global here? Not sure what PARTITION_SIZE does, or why would it be different on RoCM

tjtanaa replied:
@tlrmchlsmth This line is defined to tell the compiler that the `PARTITION_SIZE` within the test scope `test_paged_attention` function that `PARTITION_SIZE` should be from the global variable. This is needed after we defined a line https://github.com/vllm-project/vllm/blob/9501972249ca7dca0704bceb4308163f30999a6d/tests/kernels/test_attention.py#L217C36-L219C49
to reassign `PARTITION_SIZE` to have the value of `PARTITION_SIZE_ROCM`, which causes the compiler to think that `PARTITION_SIZE` is a local variable.

`PARTITION_SIZE_ROCM` is a performance-tuned hyperparameter for ROCm custom paged attention. That's why it is different from the `PARTITION_SIZE` on other platform.

tjtanaa replied:
@tlrmchlsmth Anything that I should update to merge this PR?

hongxiayang replied:
> @tlrmchlsmth Anything that I should update to merge this PR?

@tjtanaa Let's change the places where int64_t is needed to avoid overflow. You can do  static_cast<int64_t> when needed. Then it should be able to go.

tlrmchlsmth replied:
@tjtanaa and @hongxiayang are there any unit tests for this kernel that we could take from the RoCM fork? If possible that would be very nice to have

tjtanaa replied:
@tlrmchlsmth The unit tests for this kernel in this PR is the same as the one in ROCm/vllm fork. It is an optimized implementation of the original ROCm Custom Paged Attention. So, the new and older version is sharing the same unit tests, which ensure that the kernel still retains its original functionality.

tlrmchlsmth commented:
I'll take a look tomorrow

tlrmchlsmth commented on vllm/attention/ops/nki_flash_attn.py:
Could you revert changes to files that aren't relevant to for ROCm attention?

tjtanaa replied:
The pre-commit linter hook does not like the spelling. Either we should fix the spelling or update the pyproject.toml. Which one would you prefer?

And the code being formatted is based on the pre-commit hook. Should we by-pass the pre-commit hook lint check?

```
 git commit -s -m "revert changes to unrelated files as per requested"
yapf.....................................................................Passed
ruff.....................................................................Passed
codespell................................................................Failed
- hook id: codespell
- exit code: 65

vllm/attention/ops/nki_flash_attn.py:116: arthimetic ==> arithmetic
vllm/lora/punica_wrapper/punica_base.py:150: indicies ==> indices

isort....................................................................Passed
clang-format.........................................(no files to check)Skipped
PyMarkdown...........................................(no files to check)Skipped
Lint GitHub Actions workflow files...................(no files to check)Skipped
Run mypy for local Python installation...................................Passed
Lint shell scripts...................................(no files to check)Skipped
Lint PNG exports from excalidraw.....................(no files to check)Skipped
Check SPDX headers.......................................................Passed
Suggestion...............................................................Passed
- hook id: suggestion
- duration: 0s

To bypass pre-commit hooks, add --no-verify to git commit.

Check for spaces in all filenames........................................Passed
```

tjtanaa replied:
@tlrmchlsmth Yes. thank you. We have pulled the latest change in main and the these misspellings have been fixed.

tlrmchlsmth commented on vllm/lora/punica_wrapper/punica_base.py:
ditto

tjtanaa replied:
The pre-commit linter hook does not like the spelling. Either we should fix the spelling or update the pyproject.toml. Which one would you prefer?

```
 git commit -s -m "revert changes to unrelated files as per requested"
yapf.....................................................................Passed
ruff.....................................................................Passed
codespell................................................................Failed
- hook id: codespell
- exit code: 65

vllm/attention/ops/nki_flash_attn.py:116: arthimetic ==> arithmetic
vllm/lora/punica_wrapper/punica_base.py:150: indicies ==> indices

isort....................................................................Passed
clang-format.........................................(no files to check)Skipped
PyMarkdown...........................................(no files to check)Skipped
Lint GitHub Actions workflow files...................(no files to check)Skipped
Run mypy for local Python installation...................................Passed
Lint shell scripts...................................(no files to check)Skipped
Lint PNG exports from excalidraw.....................(no files to check)Skipped
Check SPDX headers.......................................................Passed
Suggestion...............................................................Passed
- hook id: suggestion
- duration: 0s

To bypass pre-commit hooks, add --no-verify to git commit.

Check for spaces in all filenames........................................Passed
```

tlrmchlsmth commented on vllm/spec_decode/spec_decode_worker.py:
ditto

tjtanaa replied:
The code is formated based on the pre-commit hook. Should we by-pass the pre-commit hook lint check?

tlrmchlsmth replied:
I think this diff will also disappear when you merge in latest main.

tjtanaa replied:
@tlrmchlsmth Yes. thank you. We have pulled the latest change in main and the diff disappeared.

tlrmchlsmth commented on csrc/rocm/attention.cu:
```diff
@@ -63,23 +66,34 @@ typedef struct _B16x8 {
-__device__ __forceinline__ void store(T value, T* addr) {
-  addr[0] = value;
+__device__ __forceinline__ _B16x8 load_ntmprl_16Byte(const _B16x8* addr) {
+  auto addr_alias = reinterpret_cast<const float*>(addr);
+  auto dat0 = loadnt(addr_alias);
+  auto dat1 = loadnt(addr_alias + 1);
+  auto dat2 = loadnt(addr_alias + 2);
+  auto dat3 = loadnt(addr_alias + 3);
+  auto res = make_float4(dat0, dat1, dat2, dat3);
+  return *reinterpret_cast<_B16x8*>(&res);
 }
+///////////////////////////////////
 
 template <typename T, int absz, int cbid, int blgp>
-__device__ __forceinline__ floatx4 gcn_mfma_instr(const _B16x4& inpA,
-                                                  const _B16x4& inpB,
-                                                  const floatx4& inpC) {
+__device__ __forceinline__ floatx4 gcn_mfma4x4x4_instr(const _B16x4& inpA,
+                                                       const _B16x4& inpB,
+                                                       const floatx4& inpC) {
```
Does the `gcn_mfma4x4x4` instruction impose any hardware or software versioning requirements that this kernel didn't have before?

tjtanaa replied:
Given that vLLM officially supports MI200 and above, `gcn_mfma4x4x4` does not impose any hardware or software versioning requirements.

This kernel has been implemented in ROCm/vllm https://github.com/ROCm/vllm/pull/372. It is a kernel that has the same feature parity with the old kernels. So we think there is no need to implement fallbacks. The kernel has been unit tested in `tests/kernels/test_attentions.py` and is rigorously tested to pass all the `tests/models/decoder_only` (for any models that are supported in ROCm)

tlrmchlsmth commented on csrc/rocm/attention.cu:
```diff
@@ -192,15 +247,597 @@ __device__ __forceinline__ _B16x8 scaled_convert_b8x8(const _B8x8 input,
+}
+
+__device__ __forceinline__ floatx4 to_float_fp8x4(const _B8x4& inp) {
+  #if defined(__gfx90a__)
+  float4 f32x4 = vllm::fp8::vec_conversion<float4, uint32_t>(
+      *reinterpret_cast<const uint32_t*>(&inp));
+  return *reinterpret_cast<floatx4*>(&f32x4);
+  #else  // MI3xx+ optimized builtins
```
I was a little confused by this since I didn't initially catch that `__gfx90a__` was the least-capable target supported by this kernel. Could you add a comment explaining this fallback case and when/why it's used?

poyenc replied:
@tlrmchlsmth From MI300+ platforms, we have `v_cvt_pk_f32_fp8` instruction to convert 2 packed fp8 to 2 packed fp32 values. However, in MI200 platforms, we only have `v_cvt_f32_fp8` which is only capable of converting fp8 values separately. So we added `#else` case for fewer instructions (=2) in MI300+, and fallback to `#if` case for other platforms.

tjtanaa replied:
@tlrmchlsmth  Thank you for your feedback. For clarity, we have add a comment to explain the fallback in the `attention.cu`

```diff
+ 250 __device__ __forceinline__ floatx4 to_float_fp8x4(const _B8x4& inp) {
+ 251   // From MI300+ platforms, we have v_cvt_pk_f32_fp8 instruction
+ 252   // to convert 2 packed fp8 to 2 packed fp32 values.
+ 253   // However, in MI200 platforms, we only have v_cvt_f32_fp8
+ 254   // to convert fp8 values individually. So we added
+ 255   // #else case for fewer instructions (# inst=2) in MI300+,
+ 256   // and fallback to
+ 257   // #if case for other platforms (# inst=4).
+ 258   #if defined(__gfx90a__)
+ 259   float4 f32x4 = vllm::fp8::vec_conversion<float4, uint32_t>(
+ 260       *reinterpret_cast<const uint32_t*>(&inp));
+ 261   return *reinterpret_cast<floatx4*>(&f32x4);
+ 262   #else  // MI3xx+ optimized builtins
+ 263   const auto f0 = __builtin_amdgcn_cvt_pk_f32_fp8(inp, false);
+ 264   const auto f1 = __builtin_amdgcn_cvt_pk_f32_fp8(inp, true);
+ 265   floatx4 ret;
+ 266   ret[0] = f0[0];
+ 267   ret[1] = f0[1];
+ 268   ret[2] = f1[0];
+ 269   ret[3] = f1[1];
+ 270   return ret;
+ 271   #endif
+ 272 }
```

tlrmchlsmth replied:
Thanks for the comment, it helps a lot

tlrmchlsmth commented on csrc/rocm/attention.cu:
```diff
@@ -192,15 +247,597 @@ __device__ __forceinline__ _B16x8 scaled_convert_b8x8(const _B8x8 input,
   }
 }
 
+template <typename T>
+__device__ __forceinline__ _B16x8
+scaled_convert_b8x8_custom(const _B8x8 input, const float scale) {
```
Please write short comments describing the functionality of these various conversion utilities

tjtanaa replied:
@poyenc could you help to comment on this. Thank you.

poyenc replied:
@tjtanaa since this function is no longer used. can we delete it?

tjtanaa replied:
@poyenc Sure. Let me clean it up. Thank you for your feedback.

tjtanaa replied:
@tlrmchlsmth Thank you. We have removed the function.

tlrmchlsmth commented on csrc/rocm/attention.cu:
```diff
@@ -192,15 +247,597 @@ __device__ __forceinline__ _B16x8 scaled_convert_b8x8(const _B8x8 input,
+          vllm::Fp8KVCacheDataType KV_DTYPE, typename OUTT, int BLOCK_SIZE,
+          int HEAD_SIZE, int NUM_THREADS, bool ALIBI_ENABLED, int GQA_RATIO>
+__global__
+__launch_bounds__(NUM_THREADS, 5) void paged_attention_ll4mi_QKV_mfma16_kernel(
+    const scalar_t* __restrict__ q,       // [num_seqs, num_heads, head_size]
+    const cache_t* __restrict__ k_cache,  // [num_blocks, num_kv_heads,
+                                          // head_size/x, block_size, x]
+    const cache_t* __restrict__ v_cache,  // [num_blocks, num_kv_heads,
+                                          // head_size, block_size]
+    const int num_kv_heads, const float scale,
+    const int* __restrict__ block_tables,  // [num_seqs, max_num_blocks_per_seq]
+    const int* __restrict__ context_lens,  // [num_seqs]
+    const int max_num_blocks_per_seq,
+    const float* __restrict__ alibi_slopes,  // [num_heads]
+    const int q_stride, const int kv_block_stride, const int kv_head_stride,
+    float* __restrict__ exp_sums,  // [num_seqs, num_heads, max_num_partitions]
+    float* __restrict__ max_logits,  // [num_seqs, num_heads,
+                                     // max_num_partitions]
+    scalar_t* __restrict__ out,    // [num_seqs, num_heads, max_num_partitions,
+                                   // head_size]
+    OUTT* __restrict__ final_out,  // [num_seqs, num_heads, head_size]
+    int max_ctx_blocks, const float* k_scale, const float* v_scale) {
```
Could you clean up the formatting for these tensor shape annotations? You can turn off clang-format if that helps

tjtanaa replied:
@tlrmchlsmth Thank you. We have cleaned up the tensor shape annotations.

```diff
+ 322 __launch_bounds__(NUM_THREADS, 5) void paged_attention_ll4mi_QKV_mfma16_kernel(
+ 323    const scalar_t* __restrict__ q,         // [num_seqs, num_heads, head_size]
+ 324    const cache_t* __restrict__ k_cache,    // [num_blocks, num_kv_heads, head_size/x, block_size, x]
+ 325    const cache_t* __restrict__ v_cache,    // [num_blocks, num_kv_heads, head_size, block_size]
+ 326    const int num_kv_heads,   
+ 327    const float scale,    
+ 328    const int* __restrict__ block_tables,   // [num_seqs, max_num_blocks_per_seq]
+ 329    const int* __restrict__ context_lens,   // [num_seqs]
+ 330    const int max_num_blocks_per_seq,
+ 331    const float* __restrict__ alibi_slopes, // [num_heads]
+ 332    const int q_stride,
+ 333    const int kv_block_stride,
+ 334    const int kv_head_stride,
+ 335    float* __restrict__ exp_sums,           // [num_seqs, num_heads, max_num_partitions]
+ 336    float* __restrict__ max_logits,         // [num_seqs, num_heads, max_num_partitions]
+ 337    scalar_t* __restrict__ out,             // [num_seqs, num_heads, max_num_partitions, head_size]
+ 338    OUTT* __restrict__ final_out,           // [num_seqs, num_heads, head_size]
+ 339    int max_ctx_blocks, const float* k_scale, const float* v_scale) {
 ```

tlrmchlsmth commented on csrc/rocm/attention.cu:
```diff
@@ -192,15 +247,597 @@ __device__ __forceinline__ _B16x8 scaled_convert_b8x8(const _B8x8 input,
+  // for QK mfma16x16, layout is QHead/Tokenx16 across every 16 lanes, 16 Bytes
+  // HeadElements in each lane, 4x16B HeadElements across 4 rows of warp
+  constexpr int ROWS_PER_WARP =
+      WARP_SIZE / 16;  // rows refers to 16 lanes; refer dpp terminology
```
QQ: what is dpp terminology?

tjtanaa replied:
DPP means Data Parallel Processing

tjtanaa replied:
@tlrmchlsmth Thank you. We have added the fullname instead of the acronym

```diff
+ 371 // for QK mfma16x16, layout is QHead/Tokenx16 across every 16 lanes, 16 Bytes
+ 372 // HeadElements in each lane, 4x16B HeadElements across 4 rows of warp
+ 373 constexpr int ROWS_PER_WARP =
+ 374     WARP_SIZE / 16;  // rows refers to 16 lanes; refer DDP (Data Parallel
+ 375                       // Processing) terminology
```

tlrmchlsmth commented on csrc/rocm/attention.cu:
```diff
@@ -192,15 +247,597 @@ __device__ __forceinline__ _B16x8 scaled_convert_b8x8(const _B8x8 input,
+    }
+  }
+
+  // set to true to enable non temporal kv loads: has some benefit in very high
+  // batch size cases
+  constexpr bool NT_KV_LOAD = false;
```
Do you know what sizes this starts to make sense to do? And does this help for large prefills as well? Or only large batch sized decodes? Wondering if we could simplify the code by removing this case

tjtanaa replied:
@poyenc could you help to comment on this. Thank you.

poyenc replied:
@tlrmchlsmth I think we can just remove it

tjtanaa replied:
@tlrmchlsmth Thank you for your feedback. We have removed this from the code.

tlrmchlsmth commented on csrc/rocm/attention.cu:
```diff
@@ -192,15 +247,597 @@ __device__ __forceinline__ _B16x8 scaled_convert_b8x8(const _B8x8 input,
+  for (int vtoken_depth = 0; vtoken_depth < VTLOOP; vtoken_depth++) {
+    for (int vblock_depth = 0; vblock_depth < VBLOCKS_PER_LANE;
+         vblock_depth++) {
+      const int vlocal_token_idx =
+          vtoken_depth * VTOKENS_PER_LANE * ROWS_PER_WARP +
+          rowid * VTOKENS_PER_LANE + vblock_depth * BLOCK_SIZE;
```
Is it possible to overflow an int32 here?

tjtanaa replied:
@poyenc could you help to comment on this, will it have other implications. Thank you.

poyenc replied:
@tlrmchlsmth it is hardly overflow an int32, because the `local_token_idx` never exceeds 256. However, the `global_token_idx` may do.

tlrmchlsmth replied:
Upon revisiting this one, this actually looks like it might have been safe.

However, I think there's a minor issue here to fix this up, because casting `partition_start_token_idx` like this probably won't have an effect, since `vlocal_token_idx` is at most 256.
```
      // Use int64_t for arithmetic to prevent overflow
      const int64_t vglobal_token_idx =
          static_cast<int64_t>(partition_start_token_idx) + vlocal_token_idx;
```

If `vglobal_token_idx` _were_ too large to fit in 32 bits, I think we'd likely overflow earlier on line 357:
```
  const int partition_start_token_idx =
      partition_idx * T_PAR_SIZE;  // partition_size;
```

But that actually looks safe. @poyenc and @tjtanaa if you agree this is safe, we can switch back to int32, otherwise I think we should change `partition_start_token_idx` to an int64

poyenc replied:
@tjtanaa I think `partition_start_token_idx` itself may overflow. So we should change to `int64_t` as well.

tlrmchlsmth commented on csrc/rocm/attention.cu:
```diff
@@ -192,15 +247,597 @@ __device__ __forceinline__ _B16x8 scaled_convert_b8x8(const _B8x8 input,
+  // write out partition max_logits and exp_sum
+  if (threadIdx.x < GQA_RATIO) {
+    const int qhead_idx = lane16id;
+    const int offset = seq_idx * total_num_heads * max_num_partitions +
+                       (wg_start_head_idx + qhead_idx) * max_num_partitions +
+                       partition_idx;
```
ditto: possible to overflow an int32 here? This one looks a little more concerning so we should consider doing the arithmetic and storing `offset` as an int64_t

tjtanaa replied:
@poyenc could you help to comment on this, will it have other implications. Thank you.

poyenc replied:
@tlrmchlsmth I agree. `offset` may overflow the int32.

tjtanaa replied:
@tlrmchlsmth  This has been addressed. Please take a look.

```diff
+ 697  // write out partition max_logits and exp_sum
+ 698  if (threadIdx.x < GQA_RATIO) {
+ 699    const int qhead_idx = lane16id;
+ 700    const int64_t offset = static_cast<int64_t>(seq_idx) *
+ 701                               static_cast<int64_t>(total_num_heads) *
+ 702                               static_cast<int64_t>(max_num_partitions) +
+ 703                           (static_cast<int64_t>(wg_start_head_idx) +
+ 704                            static_cast<int64_t>(qhead_idx)) *
+ 705                               static_cast<int64_t>(max_num_partitions) +
+ 706                           static_cast<int64_t>(partition_idx);
+ 707    max_logits[offset] = partition_qk_max;
+ 708    exp_sums[offset] = partition_exp_sum;
+ 709  }
  ```

tlrmchlsmth commented on csrc/rocm/attention.cu:
```diff
@@ -192,15 +247,597 @@ __device__ __forceinline__ _B16x8 scaled_convert_b8x8(const _B8x8 input,
+      }
+
+      const int hsz_maxp_mult = HEAD_SIZE * max_num_partitions;
+      scalar_t* out_ptr = out + seq_idx * total_num_heads * hsz_maxp_mult +
+                          partition_idx * HEAD_SIZE;
```
ditto: likely use an int64_t for `seq_idx * total_num_heads * hsz_maxp_mult`

tjtanaa replied:
@tlrmchlsmth This has been addressed. Please take a look.

```diff

+ 798      const int64_t hsz_maxp_mult =
+ 799          static_cast<int64_t>(HEAD_SIZE * max_num_partitions);
+ 800      scalar_t* out_ptr = out + seq_idx * total_num_heads * hsz_maxp_mult +
+ 801                          partition_idx * HEAD_SIZE;
```

tlrmchlsmth commented on csrc/rocm/attention.cu:
```diff
@@ -192,15 +247,597 @@ __device__ __forceinline__ _B16x8 scaled_convert_b8x8(const _B8x8 input,
+        const int local_head_idx = 4 * h + rowid;
+        if (local_head_idx < GQA_RATIO) {
+          const int out_head_idx = wg_start_head_idx + local_head_idx;
+          scalar_t* out_ptr2 = out_ptr + out_head_idx * hsz_maxp_mult;
```
and `out_head_idx * hsz_maxp_mult`

tjtanaa replied:
@tlrmchlsmth This has been addressed. Please take a look.

```diff
+ 802      for (int h = 0; h < GQA_RATIO4; h++) {
+ 803        const int local_head_idx = 4 * h + rowid;
+ 804        if (local_head_idx < GQA_RATIO) {
+ 805          const int64_t out_head_idx =
+ 806              static_cast<int64_t>(wg_start_head_idx + local_head_idx);
+ 807          scalar_t* out_ptr2 = out_ptr + out_head_idx * hsz_maxp_mult;
+ 808          scalar_t* out_ptr3 = out_ptr2 + head_elem_idx;
+ 809          _B16x8* out_ptr_B16x8 = reinterpret_cast<_B16x8*>(out_ptr3);
+ 810          *out_ptr_B16x8 = vout[h];
+ 811       }
+ 812      }
```

tlrmchlsmth commented on csrc/rocm/attention.cu:
```diff
@@ -840,39 +1483,73 @@ __launch_bounds__(NUM_THREADS) void paged_attention_ll4mi_reduce_kernel(
-      out + seq_idx * num_heads * HEAD_SIZE + head_idx * HEAD_SIZE;
-  out_ptr[threadIdx.x] = from_float<scalar_t>(acc);
+
+  OUTT* out_ptr = out + seq_idx * num_heads * HEAD_SIZE + head_idx * HEAD_SIZE;
```
I think this arithmetic should be in int64_t as well

tjtanaa replied:
@tlrmchlsmth This has been addressed. Please take a look.

```diff

+ 798      const int64_t hsz_maxp_mult =
+ 799          static_cast<int64_t>(HEAD_SIZE * max_num_partitions);
+ 800      scalar_t* out_ptr = out + seq_idx * total_num_heads * hsz_maxp_mult +
+ 801                          partition_idx * HEAD_SIZE;
```

tlrmchlsmth reviewed (commented):
Thank you for this contribution! The performance improvements look very nice and would be great to get into mainline vLLM. I took a pass through the kernel but will likely need to spend more time to understand it better. Are there any unit tests for this kernel? Especially given how dense this PR is I would feel a lot more confident landing it if it had better testing.

vllmellm committed:
address PR feedback and improve code quality:
- Clean up code and improve formatting
- Modify data types of specific variable

Signed-off-by: vllmellm <vllm.ellm@embeddedllm.com>

vllmellm committed:
change datatype in attention.cu based on reviewer; and revert some file changes

Signed-off-by: vllmellm <vllm.ellm@embeddedllm.com>

poyenc committed:
add comment to the attention.cu

Signed-off-by: poyenc <poyen.chen@amd.com>

tjtanaa committed:
convert datatype to int64 to prevent overflow as per reviewed

Signed-off-by: tjtanaa <tunjian.tan@embeddedllm.com>

vllmellm committed:
Merge remote-tracking branch 'origin/main' into port-rocm-cpa-credit

Signed-off-by: vllmellm <vllm.ellm@embeddedllm.com>

tlrmchlsmth commented on .buildkite/run-amd-test.sh:
```diff
@@ -77,7 +77,6 @@ echo "Commands:$commands"
 #ignore certain kernels tests
 if [[ $commands == *" kernels "* ]]; then
   commands="${commands} \
-  --ignore=kernels/test_attention.py \
```
Will be happy to enable this test in buildkite if it's green!

tjtanaa replied:
We ran this test on MI300X locally and it is all green.

tlrmchlsmth commented on csrc/rocm/attention.cu:
```diff
@@ -63,23 +66,34 @@ typedef struct _B16x8 {
-__device__ __forceinline__ T load(T* addr) {
-  return addr[0];
+__device__ __forceinline__ T loadnt(T* addr) {
+  return __builtin_nontemporal_load(addr);
 }
 
-template <typename T>
-__device__ __forceinline__ void store(T value, T* addr) {
-  addr[0] = value;
+__device__ __forceinline__ _B16x8 load_ntmprl_16Byte(const _B16x8* addr) {
+  auto addr_alias = reinterpret_cast<const float*>(addr);
+  auto dat0 = loadnt(addr_alias);
+  auto dat1 = loadnt(addr_alias + 1);
+  auto dat2 = loadnt(addr_alias + 2);
+  auto dat3 = loadnt(addr_alias + 3);
+  auto res = make_float4(dat0, dat1, dat2, dat3);
+  return *reinterpret_cast<_B16x8*>(&res);
 }
+///////////////////////////////////
```
Is this used anywhere? One general comment is that we should not try to add unused/untested/unexercised code to the codebase, so could you remove any other functions or types that aren't being used in this file?

tlrmchlsmth commented on csrc/rocm/attention.cu:
```diff
@@ -102,6 +131,23 @@ __device__ __forceinline__ float to_float(const T& inp) {
   }
 }
 
+template <typename T>
+__device__ __forceinline__ float to_float_b16(const bit16_t& inp) {
+  union tmpcvt {
+    bit16_t u;
+    _Float16 f;
+    __hip_bfloat16 b;
+  } t16;
+  t16.u = inp;
+  if constexpr (std::is_same<T, _Float16>::value) {
+    return (float)t16.f;
+  } else if constexpr (std::is_same<T, __hip_bfloat16>::value) {
+    return __bfloat162float(t16.b);
+  } else {
+    static_assert(false, "unsupported 16b dtype");
+  }
+}
```
ditto, looks unused

tlrmchlsmth commented on csrc/rocm/attention.cu:
```diff
@@ -396,164 +1029,153 @@ __global__ __launch_bounds__(NUM_THREADS) void paged_attention_ll4mi_QKV_kernel(
           for (int d = 0; d < BLOCK_SIZE / 8; d++) {
-            // Vlocalb8[h][b * BLOCK_SIZE / 8 + d] = v_ptrh8be[d];
-            const _B8x8 Vlocalb8 = v_ptrh8be[d];
-            Vlocal[h][b * BLOCK_SIZE / 8 + d] =
-                scaled_convert_b8x8<scalar_t, KV_DTYPE>(Vlocalb8, *v_scale_ptr);
```
Similar: please delete `scaled_convert_b8x8` now that it's not being used anymore

hongxiayang commented:
cc @houseroad This is an important feature that should be merged asap.

tjtanaa committed:
remove unused functions from attention.cu

Signed-off-by: tjtanaa <tunjian.tan@embeddedllm.com>

tlrmchlsmth commented on csrc/rocm/attention.cu:
```diff
@@ -150,75 +167,626 @@ __device__ __forceinline__ _B16x4 addx4(const _B16x4& inp1,
+      const int vlocal_token_idx =
+          vtoken_depth * VTOKENS_PER_LANE * ROWS_PER_WARP +
+          rowid * VTOKENS_PER_LANE + vblock_depth * BLOCK_SIZE;
+      // Use int64_t for arithmetic to prevent overflow
+      const int64_t vglobal_token_idx =
+          static_cast<int64_t>(partition_start_token_idx) + vlocal_token_idx;
```
Small nit that this should would be better as the following, per conversation above
```suggestion
      // Safe to use an int32_t here assuming we are working with < 2 billion tokens
      const int32_t vglobal_token_idx = partition_start_token_idx + vlocal_token_idx;
```

tjtanaa replied:
@tlrmchlsmth We have done updating this back to `int32`, but we are using `int` type to represent `int32_t` type. Is that ok?

tlrmchlsmth replied:
Yep, I think that's fine especially as the rest of the file uses `int`

tlrmchlsmth reviewed (approved):
LGTM now, thanks for the contribution! Please merge latest main as there is a conflict.

tjtanaa committed:
add comment to line 471 in attention.cu

Signed-off-by: tjtanaa <tunjian.tan@embeddedllm.com>

tjtanaa committed:
merge changes from main

Signed-off-by: tjtanaa <tunjian.tan@embeddedllm.com>

tjtanaa committed:
fix attention.cu bug as main removed some header files

Signed-off-by: tjtanaa <tunjian.tan@embeddedllm.com>

tjtanaa commented:
> LGTM now, thanks for the contribution! Please merge latest main as there is a conflict.

@tlrmchlsmth We have resolved the merge conflict. Thank you.

tlrmchlsmth commented:
@tjtanaa could you take a look at the pre-commit? It's failing as well

tjtanaa committed:
fix linter

Signed-off-by: tjtanaa <tunjian.tan@embeddedllm.com>

tjtanaa committed:
fix linter

Signed-off-by: tjtanaa <tunjian.tan@embeddedllm.com>
